Add Avatto N-TRV16 TRV (_TZE204_vjpaih9f)

DerGary opened the pull request:
## Proposed change
Adds support for Avatto N-TRV16 Tuya TS0601 Thermostat
<img width="661" height="1064" alt="image" src="https://github.com/user-attachments/assets/40296010-4449-44e2-b251-45e298f15e1d" />



## Additional information
I used the DP from zigbee2mqtt as a baseline: https://github.com/Koenkk/zigbee-herdsman-converters/blob/f963b918211305f63b7a35c3f847b55cc926cd83/src/devices/tuya.ts#L20684-L20750


## Device diagnostics
[zha-01J9PF144WNVNQYSX5HWT0ZHEW-_TZE204_vjpaih9f TS0601-ce9a45da3e81d362d04138e0044a53f4.json](https://github.com/user-attachments/files/23697348/zha-01J9PF144WNVNQYSX5HWT0ZHEW-_TZE204_vjpaih9f.TS0601-ce9a45da3e81d362d04138e0044a53f4.json)


## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [x] The changes are tested and work correctly (I was unable to trigger the window detection)
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [ ] Tests have been added to verify that the new code works
- [x] Device diagnostics data has been attached



Closes #4471

DerGary committed:
adds support for Avatto N-TRV16 Tuya TS0601 Thermostat

DerGary committed:
Formats code

MrSplit commented:
Hi @DerGary, I'm using your quirk and it works very well for me, thanks! 
Is there any update, any new version? 
Is this included in the official ZHA already? 
Thanks!

DerGary commented:
Hi @MrSplit No there IS No Update yet. I thought that someone of the zigpy codeowners would review the PR and comment If there are Changes necessary to get IT merged but there hast been No interaction at all. Would be nice If someone would review the PR.

mircoboschi commented:
I'm stuck with zigbee2mqtt, would love to see this merged.

DerGary commented:
@mircoboschi using zigbee2mqtt isn't bad. But if you really want to use zha you can copy the custom quirk file to your home assistant installation. The target path is config/zha_quirks. Restart ZHA and you should be good to go.

renaudguerin commented:
@DerGary 

Hey, just continuing our conversation from #4471 here. During my experiments with this quirk I saw this in the logs when I tried turning the TRVs on/off in the climate entity UI :
```
Logger: zigpy.zcl
Source: components/zha/climate.py:217
First occurred: 25 January 2026 at 22:37:10 (28 occurrences)
Last logged: 25 January 2026 at 23:57:07

[0xA5C8:1:0xef00] no MCU command for data TuyaClusterData(endpoint_id=1, cluster_name='thermostat', cluster_attr='system_mode', attr_value=<SystemMode.Off: 0>, expect_reply=False)
[0xA5C8:1:0xef00] no MCU command for data TuyaClusterData(endpoint_id=1, cluster_name='thermostat', cluster_attr='system_mode', attr_value=<SystemMode.Heat: 4>, expect_reply=False)
[0x52AF:1:0xef00] No cluster_dp found for 1, system_mode
[0x52AF:1:0xef00] no MCU command for data TuyaClusterData(endpoint_id=1, cluster_name='thermostat', cluster_attr='system_mode', attr_value=<SystemMode.Off: 0>, expect_reply=False)
[0x52AF:1:0xef00] no MCU command for data TuyaClusterData(endpoint_id=1, cluster_name='thermostat', cluster_attr='system_mode', attr_value=<SystemMode.Heat: 4>, expect_reply=False)
```

Below is what Claude had to say about this (I had previously fed it the quirk source code, link to the reference z2m implementation, Zigbee device diagnostics etc). I don't know enough about ZHA quirks to tell if this makes sense or not. Apologies if the below is just AI slop, but hopefully it's legit and helpful :)

---

**The quirk doesn't map `system_mode` to a Tuya datapoint!**
Looking at the Z2M converter:
```javascript
[2, "preset", tuya.valueConverterBasic.lookup({
    manual: tuya.enum(0),
    schedule: tuya.enum(1),
    ...
    off: tuya.enum(6),  // ← "off" is preset mode 6!
})],
```
In Z2M, turning the TRV "off" sends `dp_id=2, value=6` (preset=off). But in the ZHA quirk, it's exposed as `preset_mode` enum, and when HA's climate entity tries to turn off via `system_mode`, there's no datapoint mapping — the command goes nowhere.
This means your "Off" commands from HA might not be reaching the device at all, or are being ignored.
The quirk needs to either:

1. Map `system_mode` to `preset_mode` internally (Off → preset 6)
2. Or expose the climate entity differently

Here's a potential fix to add to the quirk — but this requires modifying the custom quirk file:
```python
# In the TuyaThermostat class, add system_mode handling
async def write_attributes(self, attributes, manufacturer=None):
    """Handle system_mode by converting to preset_mode."""
    if "system_mode" in attributes:
        mode = attributes.pop("system_mode")
        if mode == Thermostat.SystemMode.Off:
            # Convert to preset_mode "off" (value 6)
            attributes["preset_mode"] = PresetMode.Off
    return await super().write_attributes(attributes, manufacturer)
  ```

DerGary commented:
@renaudguerin the log suggests that we need to convert from home assistant preset mode to tuya preset mode, which makes sense. I think what claude suggested is rubbish in this case. I looked at another zigpy quirk for a tuya trv and they implemented it like this:

```
.tuya_dp_multi(
        dp_id=2,
        attribute_mapping=[
            DPToAttributeMapping(
                ep_attribute=TuyaThermostatV2.ep_attribute,
                attribute_name=TuyaThermostatV2.AttributeDefs.system_mode.name,
                converter=lambda x: {
                    TuyaPresetMode.Auto: Thermostat.SystemMode.Auto,
                    TuyaPresetMode.Eco: Thermostat.SystemMode.Auto,
                    TuyaPresetMode.Heat: Thermostat.SystemMode.Heat,
                    TuyaPresetMode.Off: Thermostat.SystemMode.Off,
                }[x],
                dp_converter=lambda x: {
                    Thermostat.SystemMode.Auto: TuyaPresetMode.Auto,
                    Thermostat.SystemMode.Heat: TuyaPresetMode.Heat,
                    Thermostat.SystemMode.Off: TuyaPresetMode.Off,
                }[x],
            ),
            DPToAttributeMapping(
                ep_attribute=TuyaMCUCluster.ep_attribute,
                attribute_name="preset_mode",
            ),
        ],
    )
```

I don't fully understand that code but the necessity to convert the mode seems to be legit.

I added this to my todo list and will look deeper into it when I got some time to spare.

renaudguerin commented:
@DerGary Based on your comment, I've just made [a new PR](https://github.com/DerGary/zha-device-handlers/pull/1) against your branch. It improves system_mode<->preset_mode mapping and the climate card now works correctly for me.

Feel free to merge into this one if you like.

schabau commented:
Hello,

Am I misunderstanding this thermostat, which has been driving me crazy for days, or is there a problem with your Quirks?

I've tried the Quirks from @DerGary and also the Quirks from @renaudguerin. The result is always the same, except that the aforementioned error is fixed in @renaudguerin's Quirks.

The presets are important to me, and apparently, I can't set them in Home Assistant. I can set the respective temperatures for each mode (Eco, Comfort, FrostProtection, Holiday) in Home Assistant, and when I switch to the respective mode on the thermostat itself, this temperature is used, and the mode is also displayed in select.thermostat_bad_voreinstellung in Home Assistant. However, changing select.thermostat_bad_voreinstellung in Home Assistant has no effect on the thermostat. The presets (Eco, Comfort, FrostProtection, Holiday) are unusable in Home Assistant. From Home Assistant, I can only switch the thermostat on and off and change the temperatures, but I can't select a preset.

Best regards,
Detlef Paschke

schabau commented:
I just ran a test with a Z2M virtual machine that I have here for testing purposes.

Aside from the fact that Z2M also allows you to configure the schedule, which isn't really necessary with HA, all preset modes can be selected and are applied directly to the thermostat.

Best regards,
Detlef Paschke

jancernoch-cz commented:
Hello,

I am testing your quirk and everything work smooth (except the presets, which are not necessary with HA as mentioned before).

Just one small fix, the thermostat range is from 5 to 35 degrees. I am testing and it should be enough to add those attributes into _CONSTANT_ATTRIBUTES:

`
Thermostat.AttributeDefs.abs_min_heat_setpoint_limit.id: 500,  # 5°C
Thermostat.AttributeDefs.abs_max_heat_setpoint_limit.id: 3500,  # 35°C
`

Thank you so much for great work and I would be really glad when merged.
Jan Cernoch

Commit:
Add Avatto N-TRV16 TRV (_TZE204_vjpaih9f)

Adds a v2 quirk for the AVATTO N-TRV16 Tuya thermostatic radiator
valve, exposing preset mode, running state, heating setpoint, local
temperature (+calibration), battery, child lock, window
detection/open, frost protection, and eco/comfort/holiday/frost
temperatures.

Incorporates fixes contributed via review on the upstream PR:
- dp 2 now maps to both system_mode (climate on/off) and the raw
  preset_mode attribute via tuya_dp_multi, so turning the thermostat
  on/off from the HA climate card actually reaches the device
  (previously only preset_mode was mapped, so system_mode writes were
  silently dropped). Contributed by @renaudguerin.
- the dp_converter for dp 2 is now a single top-level callback
  accepting one positional value per mapped attribute (system_mode,
  preset_mode), matching the framework's calling convention; the
  previous per-mapping dp_converter took only one argument and raised
  TypeError on every write to system_mode or preset_mode.
- abs_min/max_heat_setpoint_limit are now set to 5-35C to match the
  device's real range. Contributed by @jancernoch-cz.
- fixed the missing .py extension, which meant the quirk was never
  imported/registered at all.
- fixed a window_detection fallback_name that collided with the
  fallback_name used by other quirks sharing that translation_key.
- switched to the canonical zhaquirks.builder import path instead of
  the deprecated zigpy.quirks.v2 / zigpy.quirks.v2.homeassistant
  paths.
- added unit tests covering the dp read/write mapping.

Known limitation: selecting a preset other than on/off/auto (Eco,
Comfort, FrostProtection, Holiday) via the HA select entity does not
currently reach the device, since the dp_converter derives the raw
Tuya value from system_mode only. This mirrors the existing pattern
in tuya_trv.py and needs real hardware to fix properly.

DerGary commented:
Rebased on `dev` and pushed a batch of fixes based on the discussion in this thread:

- **On/off from the climate card**: `dp 2` now maps to both `system_mode` and the raw `preset_mode` attribute (via `tuya_dp_multi`), so turning the TRV on/off from HA's climate card actually reaches the device. Previously only `preset_mode` was mapped, so `system_mode` writes were silently dropped (`no MCU command for data ...`). Cherry-picked from @renaudguerin's fix, thank you!
- **Fixed a crash introduced by that fix**: since `dp 2` is now shared by two attributes, the framework calls the `dp_converter` with one positional value per mapped attribute. The per-mapping `dp_converter` only accepted a single argument and would raise `TypeError` on *every* write to `system_mode` or `preset_mode`. Moved it to the multi-arg top-level `dp_converter`, matching the pattern already used in `tuya_trv.py`.
- **Setpoint range**: added `abs_min_heat_setpoint_limit` / `abs_max_heat_setpoint_limit` (5-35°C) per @jancernoch-cz's report.
- **Missing `.py` extension**: the quirk file had no `.py` suffix, so it was never imported/registered by ZHA at all.
- Fixed a `fallback_name` collision on `window_detection` (CI catches this - all quirks sharing a `translation_key` must use the same `fallback_name`).
- Switched to the canonical `zhaquirks.builder` import path instead of the deprecated `zigpy.quirks.v2` / `zigpy.quirks.v2.homeassistant` paths.
- Added unit tests covering the dp read/write mapping, including a regression test for the crash above.

**Known limitation (not yet fixed)**: @schabau reported presets other than on/off (Eco, Comfort, FrostProtection, Holiday) don't apply via the HA select entity, even with the on/off fix. Root cause: the `dp_converter` derives the outgoing Tuya value from `system_mode` only, so selecting e.g. "Eco" from the select entity doesn't actually change what's sent to the device. This mirrors the existing pattern in `tuya_trv.py`'s TRV quirk (same limitation there), and I don't have this hardware to properly test a fix - happy to take a PR for it.
